### PR TITLE
Bug fixes and modest extensions

### PR DESCRIPTION
**Pull request by awruef**
Many changes to deal with running on "real" code:

 * In certain situations (macros) two AST nodes can occupy precisely the same source location. This is a problem for the re-writer that maintains a strong mapping between source locations and constraint variables, especially for whole-program analysis. The rewriter now does something conservative where two constraint variables that occupy the same source location implicitly alias.
 * In other situations, it's difficult to know what to do when re-writing structure typedefs or other kinds of typedefs. We now do something conservative where structure typedefs are left alone. 
 * There is now a bunch of defensive programming to keep the rewriting logic from trying to do rash things like rewrite a declaration that starts in one file and ends in another file.

### CODE REVIEW

File: tools/checked-c-convert/CheckedCConvert.cpp [hunk 1]
```diff
@@ -197,8 +217,16 @@ void rewrite(Rewriter &R, std::set<NewTyp *> &toRewrite, SourceManager &S,
       //       then clang will give back an invalid source range for the 
       //       return type source range. For now, check that the source
       //       range is valid. 
+      //       Additionally, a source range can be (mis) identified as 
+      //       spanning multiple files. We don't know how to re-write that,
+      //       so don't.
       SourceRange SR = UD->getReturnTypeSourceRange();
-      if(SR.isValid())
+      if(SR.isValid() && (R.getRangeSize(SR) != -1))
```
Comment (dtarditi) [inline tools/checked-c-convert/CheckedCConvert.cpp]: Nit: need a space after the 'if'

File: tools/checked-c-convert/MappingVisitor.cpp [hunk 2]
```diff
@@ -28,12 +27,6 @@ bool MappingVisitor::VisitDeclStmt(DeclStmt *S) {
       Stmt *So = NULL;
       Type *T = NULL;
       std::tie<Stmt *, Decl *, Type *>(So, D, T) = PSLtoSDT[PSL];
-      if (So != NULL) {
```
Comment (dtarditi) [inline tools/checked-c-convert/MappingVisitor.cpp]: Do you want to keep this code and put it under the Verbose flag?

File: tools/checked-c-convert/NewTyp.cpp [hunk 3]
```diff
@@ -16,51 +16,55 @@ using namespace llvm;
...
-  Constraints::EnvironmentMap env = CS.getVariables();
 
+  assert(!TL.isNull());
+  
   // Strip off function definitions from the type.
   while (!TL.isNull()) {
     QualType T = TL.getType();
     if (T->isFunctionNoProtoType() || T->isFunctionType() ||
-        T->isFunctionProtoType())
+      T->isFunctionProtoType())
```
Comment (dtarditi) [inline tools/checked-c-convert/NewTyp.cpp]: Nit: unintentional deletion of space?

File: tools/checked-c-convert/ProgramInfo.cpp [hunk 4]
```diff
@@ -46,6 +48,7 @@ bool ProgramInfo::link() {
       else if (GlobalVariableSymbol *V = dyn_cast<GlobalVariableSymbol>(U))
         vars.insert(V);
 
+    // Then, iterate over each of the function symbols F1=F,F2=F+1 found for this
```
Comment (dtarditi) [inline tools/checked-c-convert/ProgramInfo.cpp]: I found the wording of this comment confusing.

File: tools/checked-c-convert/NewTyp.h [hunk 5]
```diff
@@ -84,6 +98,12 @@ class ArrTyp : public NewTyp {
   std::string mkStr() {
     return ReferentTyp->mkStr() + "* ";
   }
+
+  virtual bool anyChanges() { return false || ReferentTyp->anyChanges(); }
```
Comment (dtarditi) [inline tools/checked-c-convert/NewTyp.h]: Why do you have "return false || ..." here?  Can you eliminate the false clause?

File: tools/checked-c-convert/NewTyp.h [hunk 6]
```diff
@@ -94,5 +114,11 @@ class WildTyp : public NewTyp {
...
```
Comment (dtarditi) [inline tools/checked-c-convert/NewTyp.h]: Same question here about why you have "false || ..." here.

File: tools/checked-c-convert/CheckedCConvert.cpp (hunk 1 above)
Comment (dtarditi) [inline tools/checked-c-convert/CheckedCConvert.cpp]: Nit: missing space after 'if'

File: tools/checked-c-convert/CheckedCConvert.cpp [hunk 7]
```diff
@@ -197,8 +217,16 @@ void rewrite(Rewriter &R, std::set<NewTyp *> &toRewrite, SourceManager &S,
...
+        R.ReplaceText(SR, N->mkStr());
+    }
+    else if (FieldDecl *FD = dyn_cast<FieldDecl>(D)) {
```
Comment (dtarditi) [inline tools/checked-c-convert/CheckedCConvert.cpp]: There's a redundant case for FieldDecl immediately after this one; I think the second one should be deleted.

**Comment by dtarditi** [comment]
Generally looks good - I've pointed out some minor nits in the code.  

The PersistentSourceLoc file needs more comments describing what it is doing.   Please clarify that the source location used for comparisons is the original textual location, not the expanded source location.
 
You might want a helper function for validating that a range can be rewritten.   Currently, you are checking that a range is valid and that getRangeSize(...) != -1.   The test that getRangeSize(...) != -1 is idiomatic,  so using a helper function would make the code clearer to read.

**Comment by awruef** [comment]
I think I've addressed the code review feedback.

**Pull request merged**